scripts/filter_multimappers.py

```diff
@@ -0,0 +1,177 @@
+    indels: int = sum(alignment_indels)
+
+    return indels
+
+
+def find_best_alignments(alignments: List[pysam.AlignedSegment]) -> List[pysam.AlignedSegment]:
+    """Find alignments with less indels.
+
+    This function creates a list of tuples with the alignment object and its
+    number of indels. Then, computes the minimum number of indels and returns
+    a list with the alignments that have no more than that minimum number of
+    indels. In addition, it updates the tag 'NH' and 'XI' to match the final
+    number of alignments kept and its identifier respectively.
+
+    Args:
+        alignments (list[pysam.AlignedSegment]): alignments with the same query name
+
+    Retrns:
+        best_alignments (list[pysam.AlignedSegment]): alignments with the less indels
+    """
+    alignment_indels = [(aln, count_indels(alignment=aln)) for aln in alignments]
+    min_indels = min(alignment_indels, key=lambda x: x[1])[1]
+    best_alignments = [indelCount[0] for indelCount in alignment_indels if indelCount[1] == min_indels]
+    
+    for i in range(len(best_alignments)):
+        best_alignments[i].set_tag('NH', len(best_alignments))
+        best_alignments[i].set_tag('XI', i)
+
+    return best_alignments
+
+
+def write_output(alignments: List[pysam.AlignedSegment]) -> None:
+    """Write the output to the standard output (stdout).
+
+    Args:
+        alignments (list[pysam.AlignedSegment]): alignments with the same query name
+    """
+    if len(alignments) == 1:
+        sys.stdout.write(alignments[0].to_string() + '\n')
+    else:
+        for alignment in alignments:
+            sys.stdout.write(alignment.to_string() + '\n')
+
+
+def main(sam_file: Path) -> None:
+    """Filter multimappers by indels count.
+
+    Args:
+        sam_file (str): Path to the input SAM file.
+    """
+    with pysam.AlignmentFile(sam_file, "r") as samfile:
+
+        # Extract header from input SAM file and write it to stdout
+        header = samfile.header
+        sys.stdout.write(str(header))
+
+        current_query = None
+        current_alignments: list[pysam.AlignedSegment] = []
+        
+        for alignment in samfile:
+ 
+            # Skip secondary and supplementary alignments
+            if alignment.is_secondary or alignment.is_supplementary:
+                continue
+
+            if current_query is None:
+                current_query = alignment.query_name
+
+            if current_query != alignment.query_name:
+                if len(current_alignments) == 1:
+                    write_output(alignments=current_alignments)
+                else:
+                    best_alignments = find_best_alignments(alignments=current_alignments)
+                    write_output(alignments=best_alignments)
+
+                # Reset variables
+                current_query = alignment.query_name
+                current_alignments = [alignment]
+            else:
+                current_alignments.append(alignment)
```

Comment:
I would switch the `if` and `else`, for two reasons:
1. The appending of alignmets of the same read/query happens _before_ their processing/filtering, so it makes sense for the code where you append those alignments to come first.
2. It is easier to parse `==` compared to `!=`, so if there is no good reason to use negatives (and in fact, there is rather another reason _not_ to use the negative, see reason above), then don't use them.